### PR TITLE
Apply YaRN scaling independently of cache length

### PR DESCRIPTION
**Pull request by ezyang**
DeepSeek originally used max_seq_len as both the configured context capability and the signal to enable extended-context YaRN. TorchTitan later began replacing max_seq_len with training.seq_len so that it only allocates the required RoPE cache.

For 4096-token training, this changes the configured maximum from 16384 to the original sequence length of 4096. The old guards then silently disable both the YaRN rotary frequencies and DeepSeek attention mscale, despite the model explicitly selecting YaRN with rope_factor=40.

Select YaRN from its explicit scaling policy and factor instead of the incidental cache allocation length. This also makes ComplexRoPE consistent with CosSinRoPE. Add regressions for the equality case covering both the rotary cache and DeepSeek softmax scale.

### CODE REVIEW

**Comment by ezyang** [comment]
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* __->__ #4117

File: torchtitan/models/deepseek_v3/model.py [hunk 1]
```diff
@@ -84,7 +84,7 @@ def __init__(self, config: Config):
         self.wo = config.wo.build()
         self.softmax_scale = self.qk_head_dim**-0.5
 
-        if config.rope.max_seq_len > config.rope.original_seq_len:
+        if config.rope.scaling == "yarn" and config.rope.rope_factor > 1.0:
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/model.py:87]: I feel I understand the issue, but I also doubt
- When we change the sequence length (from max to what input has), should we also disable YaRN and clear the rope factor? Because it doesn't make sense to me to use yarn on a short sequence.

Comment (shuhuayu) [reply torchtitan/models/deepseek_v3/model.py:87]: i think this is decoupling rope cache size `max_seq_len`, and rope scaling math (yarn plus `rope_factor > 1.0`). consider the following use case, you want to fine-tune a yarn enabled checkpoint using some short  sequenced data say 2k length. the current condition will not allow this.

File: tests/unit_tests/test_rope.py [hunk 2]
```diff
@@ -189,6 +190,47 @@ def test_forward_accepts_three_axis_positions(self):
         self.assertEqual(xk_out.shape, xk.shape)
 
 
+class TestYaRNScaling(unittest.TestCase):
+    @staticmethod
+    def _rope_config() -> ComplexRoPE.Config:
+        max_seq_len = 4096
```
Comment (shuhuayu) [inline tests/unit_tests/test_rope.py]: why not test a stronger case `max_seq_len = 2048` since one can pass this test by changing `>` to `>=` in the old condition check.

**Review by shuhuayu — APPROVED** [review]
lgtm! please fix the lint.

File: torchtitan/models/deepseek_v3/model.py (hunk 1 above)
Comment (ezyang) [reply torchtitan/models/deepseek_v3/model.py:87]: Yes. So in fact, this was discovered by trying to reproduce the March DSv3 Megatron paper, where the profile shows that yarn should be used but torchtitan is not applying yarn.

Comment (tianyu-l) [reply torchtitan/models/deepseek_v3/model.py:87]: @shuhuayu makes sense, thanks for explaining

**Review by tianyu-l — APPROVED** [review]
